packages/components/src/components/Calculator/index.tsx

```diff
@@ -0,0 +1,238 @@
+import React, { FC, ReactNode, useEffect, useReducer } from "react";
+
+import {
+  SqValue,
+  SqCalculator,
+  SqError,
+  result,
+  SqValuePath,
+} from "@quri/squiggle-lang";
+import { Env } from "@quri/squiggle-lang";
+
+import { PlaygroundSettings } from "../PlaygroundSettings.js";
+import { SqValueWithContext, valueHasContext } from "../../lib/utility.js";
+
+import ReactMarkdown from "react-markdown";
+import {
+  CalculatorState,
+  runSquiggleCode,
+  allFieldValuesAreValid,
+  allFieldResults,
+  initialState,
+} from "./calculatorReducer.js";
+
+import {
+  useViewerContext,
+  useCalculatorFns,
+} from "../SquiggleViewer/ViewerProvider.js";
+
+type Props = {
+  value: SqCalculator;
+  environment: Env;
+  settings: PlaygroundSettings;
+  valueWithContext: SqValueWithContext;
+  renderValue: (
+    value: SqValueWithContext,
+    settings: PlaygroundSettings
+  ) => ReactNode;
+};
+
+const updateFnValue = async (
+  path: SqValuePath,
+  state: CalculatorState,
+  calc: SqCalculator,
+  env: Env,
+  dispatch
+) => {
+  let finalResult: result<SqValue, SqError> | null = null;
+  if (allFieldValuesAreValid(state)) {
+    const results: SqValue[] = allFieldResults(state).map((res) => {
+      if (res && res.ok) {
+        return res.value;
+      } else {
+        throw new Error("Invalid result encountered.");
+      }
+    });
+    finalResult = calc.run(results, env);
+  } else {
+    finalResult = null;
+  }
+
+  dispatch({
+    type: "CALCULATOR_SET_FN_VALUE",
+    payload: { path, value: finalResult },
+  });
+};
+
+async function initStuff(
+  dispatch,
+  path: SqValuePath,
+  state: CalculatorState,
+  calc: SqCalculator,
+  env
+) {
+  for (const name of state.fieldNames) {
+    const field = state.fields[name];
+    const valueResult = await runSquiggleCode(field.code, env);
+    const _action = {
+      type: "CALCULATOR_SET_FIELD_VALUE",
+      payload: { name, value: valueResult, path },
+    };
+    dispatch(_action);
+  }
+  updateFnValue(path, state, calc, env, dispatch);
+}
+
+async function updateCode(
```

Comment:
I'm not sure where to put these methods now. I used to have them in the asyncReducer, which was pretty clean, though maybe not really that essential.